### ISSUE TITLE
CUDA error when using more than 2^14 angles

### ISSUE THREAD

**Issue by RendersJens** [astra-toolbox/astra-toolbox#278]
The algorithms `SIRT3D_CUDA` and `CGLS3D_CUDA` both crash when the projection geometry uses more than 2^14 angles. Here is a minimal example to recreate:
```
import numpy as np
import astra

n_angles = 2**14+1


vol_geom =  astra.create_vol_geom(1, 1, 1)
proj_geom = astra.create_proj_geom('parallel3d', 1, 1, 1, 1, np.linspace(0, np.pi, n_angles))
sino = np.random.random((1, n_angles, 1)).astype(np.float32)

sino_id = astra.data3d.create('-sino', proj_geom, sino)
rec_id = astra.data3d.create('-vol',vol_geom)

cfg = astra.creators.astra_dict('SIRT3D_CUDA')
cfg['ReconstructionDataId'] = rec_id
cfg['ProjectionDataId'] = sino_id

alg_id = astra.algorithm.create(cfg)

astra.algorithm.run(alg_id, 1)
```
The error happens on the `astra.algorithm.run` command, and it prints:
```
Error: CUDA error 1: invalid argument.
Error: Failed to allocate 1x16385x1 GPU array
Error: CUDA error 1: invalid argument.
python: ./../../cuda/3d/util3d.cu:383: bool astraCUDA3d::transferProjectionsToArray(cudaPitchedPtr, cudaArray*, const astraCUDA3d::SDimensions3D&): Assertion `err == cudaSuccess' failed.
Aborted (core dumped)

```

**Comment by snipdome** [comment]
Hi @RendersJens,
that has to do with the CUDA maximum memory dimensions of the allocable array  (which will be linked to a texturized memory representation). Specifications can be found here ( table 15, https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html )

**Comment by askorikov** [comment]
This already works in forward- and backprojection operations by automatically splitting data into blocks fitting into GPU memory and max texture size. Given this, you can try to build SIRT from FP and BP operations. We will also consider enabling this functionality in the rest of algorithms implemented in Astra where possible.

@RendersJens Just out of curiosity, do you have a real application case for more than 2^14 angles or is it more about the limits of the toolbox?

**Comment by wjp** [comment]
Astra does also already contain a python version of SIRT that should avoid this limitation. Call `astra.plugin.register(astra.plugins.SIRTPlugin)` to enable this `SIRT-PLUGIN` algorithm.

**Comment by RendersJens** [comment]
> @RendersJens Just out of curiosity, do you have a real application case for more than 2^14 angles or is it more about the limits of the toolbox?

Yes this comes from a real application, not from seeking the limits. The application is Capillary Optics, in which there are many small cone x-ray beams. Pawel Koreki gave a presentation about this in our group back in 2021 and told us that he modeled this in ASTRA but got an error. I tracked it down and that led to this bugreport. I helped him by giving a script for SIRT and BB in MATLAB just like the suggestions here, but opened the report anyway because it worked in BP and FP so it looked like a clear bug to me.

**Comment by snipdome** [comment]
Also with the case I was dealing with, it was real application. This one external company wanted to use a one line detector as a 2D detector by stacking 1D projections. For the geometry of the situation, this could be solved in astra only by considering many 2-D projections (with one side as 1) and reconstructing a 3D volume (jointly), which generated a lot of views. The bug amount of views (projections, in Jens' case), hit the max capability of the CUDA texturized array. We can say that It costed us a potential user

**Comment by askorikov** [comment]
@RendersJens, @snipdome All right, I see. This type of use cases is good to keep in mind.

@snipdome Yes, FP/BP should not currently be limited by GPU memory. It works only with `cuda3d` projector, but with all types of (3D) geometries.
